File: CPP/No264.cc
/**
 * Created by Xiaozhong on 2020/8/24.
 * Copyright (c) 2020/8/24 Xiaozhong. All rights reserved.
 */

#include "vector"
#include "iostream"

using namespace std;

class Solution {
public:
    int nthUglyNumber(int n) {
        vector<int> dp(n, 0);
        dp[0] = 1;
        int p2 = 0, p3 = 0, p5 = 0;
        for (int i = 1; i < n; i++) {
            dp[i] = min(min(dp[p2] * 2, dp[p3] * 3), dp[p5] * 5);
            if (dp[i] == dp[p2] * 2) p2++;
            if (dp[i] == dp[p3] * 3) p3++;
            if (dp[i] == dp[p5] * 5) p5++;
        }
        return dp[n - 1];
    }
};

int main() {
    Solution s;
    cout << s.nthUglyNumber(10) << endl;
}

File: Java/src/No264.java
/*
 * Copyright (c) 9/4/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Comparator;
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Set;

public class No264 {

    public int nthUglyNumber(int n) {
        if (n == 1) return 1;
        int[] factories = {2, 3, 5};
        PriorityQueue<Long> minHeap = new PriorityQueue<>((Comparator.comparingLong(o -> o)));
        minHeap.add(1L);
        Set<Long> visited = new HashSet<>();
        long uglyNumber = 0;
        while (n-- > 0){
            uglyNumber = minHeap.poll();
            for (int num : factories) {
                if (!visited.contains(num * uglyNumber)) {
                    visited.add(num * uglyNumber);
                    minHeap.add(num * uglyNumber);
                }
            }
        }
        return (int) uglyNumber;
    }

    public static void main(String[] args) {
        No264 no264 = new No264();
        System.out.println(no264.nthUglyNumber(10));
    }
}
